File: src/lib/jitter.ts
// Timing shapes that keep class-shaped events (the bell, a classroom's Wi-Fi coming back, a deploy,
// a refusal heard by 25 learners at once) from lining devices up (docs/JITTER.md). A person's first
// try is never delayed; only code that sends again without a person, or many people told "try again"
// together, gets a spread. The server says when to come back (Retry-After / retryAfterMs); these
// helpers never retry earlier than that.

/** A number in [0, 1). Injected in tests, Math.random otherwise. */
export type Rand = () => number;

/** Uniform in [lo, hi); hi <= lo gives lo. */
export function between(lo: number, hi: number, rand: Rand = Math.random): number {
  if (!(hi > lo)) return lo;
  const v = lo + rand() * (hi - lo);
  // lo + r·(hi − lo) can round up to hi itself (10000 + 0.99…·10000 is 20000 in doubles).
  return v < hi ? v : Math.max(lo, hi - Math.max(Math.abs(hi), 1) * Number.EPSILON);
}

/** AWS full jitter: uniform in [0, min(capMs, baseMs · 2^attempt)). */
export function fullJitter(attempt: number, baseMs: number, capMs: number, rand: Rand = Math.random): number {
  return between(0, Math.min(capMs, baseMs * 2 ** Math.min(Math.max(attempt, 0), 30)), rand);
}

/**
 * The wait before an automatic retry: the backoff, but never earlier than the server's hint plus up
 * to one second (which undoes the header's whole-second rounding and keeps hinted devices apart).
 */
export function retryDelay(backoffMs: number, retryAfterMs: number | null | undefined, rand: Rand = Math.random): number {
  return retryAfterMs == null ? backoffMs : Math.max(backoffMs, retryAfterMs + between(0, 1000, rand));
}

/**
 * How long a manual retry control stays disabled after a failure. Drawn once per failure and stored,
 * so a re-render never redraws it: the hint plus up to one second when the server gave one; nothing
 * for a service that is off (a timed retry would only fail again); 10–20 s for a 429 from an older
 * server without a hint; otherwise nothing.
 */
export function cooldown(
  hint: { status?: number | null; retryAfterMs?: number | null; code?: string | null },
  rand: Rand = Math.random,
): number {
  if (hint.retryAfterMs != null) return hint.retryAfterMs + between(0, 1000, rand);
  if (hint.code === 'AI_UNAVAILABLE' || hint.code === 'PROVIDER_OFF') return 0;
  return hint.status === 429 ? between(10_000, 20_000, rand) : 0;
}

/** Waits ms; an abort clears the timer and rejects with the signal's reason. */
// jitter: none — a helper that waits exactly the delay its caller drew; each caller carries its own marker
export function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  if (signal?.aborted)
    return Promise.reject(signal.reason || new DOMException('작업 확인을 중단했어요.', 'AbortError'));
  return new Promise<void>((resolve, reject) => {
    // jitter: none — the timer behind sleep(): the caller's delay, unchanged
    const timer = setTimeout(() => {
      signal?.removeEventListener('abort', cancel);
      resolve();
    }, ms);
    function cancel() {
      clearTimeout(timer);
      reject(signal?.reason || new DOMException('작업 확인을 중단했어요.', 'AbortError'));
    }
    signal?.addEventListener('abort', cancel, { once: true });
  });
}

File: server/internal/jitter/jitter.go
// Package jitter holds the timing shapes the server uses to keep class-shaped events (the bell, a
// classroom's Wi-Fi coming back, a deploy, a rate-limit window edge, lifetimes created together)
// from lining actors up (docs/JITTER.md): a uniform window (Between) and AWS full jitter (Full).
// The third shape, the server's Retry-After hint, is drawn from Between in httpx.Fail.
//
// Randomness is injected: owners hold a Rand next to their clock, tests pin it, and production uses
// Std (math/rand/v2 — goroutine-safe and seeded per process, so two instances never share a sequence).
package jitter

import (
	"math/rand/v2"
	"time"
)

// Rand returns a float64 in [0, 1).
type Rand func() float64

// Std is the production source.
func Std() float64 { return rand.Float64() }

// Between is uniform in [lo, hi); hi <= lo gives lo.
func Between(r Rand, lo, hi time.Duration) time.Duration {
	if hi <= lo {
		return lo
	}
	return lo + time.Duration(r()*float64(hi-lo))
}

// Full is AWS full jitter: uniform in [0, min(cap, base·2^attempt)). An attempt large enough to
// overflow the shift is treated as reaching the cap.
func Full(r Rand, attempt int, base, cap time.Duration) time.Duration {
	c := cap
	if attempt >= 0 && attempt < 62 {
		if v := base << attempt; v > 0 && v < cap {
			c = v
		}
	}
	return Between(r, 0, c)
}
